File: components/physics/src/test/java/org/storm/physics/Simulator.java
package org.storm.physics;

import javafx.animation.Animation;
import javafx.animation.KeyFrame;
import javafx.animation.Timeline;
import javafx.event.Event;
import javafx.util.Duration;
import lombok.Getter;
import org.storm.core.ui.Resolution;

public class Simulator {

  @Getter
  private final PhysicsEngine physicsEngine;

  private final Runnable renderHandler;

  private double lastUpdateTime;

  private final double targetFps;

  private final double fixedStepInterval;

  private double accumulator;

  public Simulator(Resolution resolution, double targetFps, Runnable renderHandler)  {
    this.physicsEngine = new ImpulseResolutionPhysicsEngine(resolution);
    this.targetFps = targetFps;
    this.fixedStepInterval = 1000000000.0 / this.targetFps;
    this.renderHandler = renderHandler;
    this.lastUpdateTime = System.nanoTime();

  }

  public void setPaused(boolean b) {
    this.physicsEngine.setPaused(b);
  }

  public boolean isPaused() {
    return this.physicsEngine.isPaused();
  }

  public void simulate() {
    KeyFrame loopFrame = new KeyFrame(Duration.millis(1000.0 / targetFps), this::run);
    Timeline timeline = new Timeline(targetFps, loopFrame);
    timeline.setAutoReverse(false);
    timeline.setCycleCount(Animation.INDEFINITE);
    timeline.play();
  }

  private void run(Event e) {
    double now = System.nanoTime();
    double elapsedFrameTime = now - lastUpdateTime;
    lastUpdateTime = now;
    accumulator += elapsedFrameTime;

    while (accumulator >= fixedStepInterval) {
      physicsEngine.update(toSeconds(lastUpdateTime), toSeconds(elapsedFrameTime));
      accumulator -= fixedStepInterval;
      lastUpdateTime += fixedStepInterval;
    }

    this.renderHandler.run();
  }

  private static double toSeconds(double nanoSeconds) {
    return nanoSeconds * 0.000000001;
  }

  private static double toMilliSeconds(double nanoSeconds) {
    return nanoSeconds / 1000000;
  }
}

File: components/physics/src/test/kotlin/org/storm/physics/visual/Simulator.kt
package org.storm.physics.visual

import javafx.animation.Animation
import javafx.animation.KeyFrame
import javafx.animation.Timeline
import javafx.event.ActionEvent
import javafx.event.Event
import javafx.util.Duration
import org.storm.core.ui.Resolution
import org.storm.physics.ImpulseResolutionPhysicsEngine
import org.storm.physics.PhysicsEngine

class Simulator(
  resolution: Resolution,
  private val targetFps: Double,
  private val renderHandler: () -> Unit
) {

  companion object {
    private fun toSeconds(nanoSeconds: Double): Double {
      return nanoSeconds * 0.000000001
    }
  }

  val physicsEngine: PhysicsEngine = ImpulseResolutionPhysicsEngine(resolution)

  private var lastUpdateTime: Double = System.nanoTime().toDouble()
  private val fixedStepInterval: Double = 1000000000.0 / this.targetFps
  private var accumulator = 0.0

  fun simulate() {
    val loopFrame = KeyFrame(Duration.millis(1000.0 / targetFps), { e: ActionEvent -> this.run(e) })
    val timeline = Timeline(targetFps, loopFrame)
    timeline.isAutoReverse = false
    timeline.cycleCount = Animation.INDEFINITE
    timeline.play()
  }

  private fun run(e: Event) {
    val now = System.nanoTime().toDouble()
    val elapsedFrameTime = now - lastUpdateTime
    lastUpdateTime = now
    accumulator += elapsedFrameTime
    while (accumulator >= fixedStepInterval) {
      physicsEngine.update(toSeconds(lastUpdateTime), toSeconds(elapsedFrameTime))
      accumulator -= fixedStepInterval
      lastUpdateTime += fixedStepInterval
    }
    renderHandler()
  }

}
